openwisp_controller/config/api/serializers.py

```diff
@@ -0,0 +1,266 @@
+import ast
+import collections
+import json
+
+from django.db import transaction
+from django.db.models import Q
+from django.utils.translation import ugettext_lazy as _
+from rest_framework import serializers
+from swapper import load_model
+
+from openwisp_users.api.mixins import FilterSerializerByOrgManaged
+from openwisp_utils.api.serializers import ValidatedModelSerializer
+
+Template = load_model('config', 'Template')
+Vpn = load_model('config', 'Vpn')
+Device = load_model('config', 'Device')
+Config = load_model('config', 'Config')
+Organization = load_model('openwisp_users', 'Organization')
+
+
+class BaseMeta:
+    read_only_fields = ['created', 'modified']
+
+
+class BaseSerializer(FilterSerializerByOrgManaged, ValidatedModelSerializer):
+    pass
+
+
+class TemplateSerializer(BaseSerializer):
+    config = serializers.JSONField(initial={})
+    tags = serializers.StringRelatedField(many=True, read_only=True)
+    default_values = serializers.JSONField(required=False, initial={})
+
+    class Meta(BaseMeta):
+        model = Template
+        fields = [
+            'id',
+            'name',
+            'tags',
+            'organization',
+            'type',
+            'backend',
+            'default',
+            'required',
+            'default_values',
+            'config',
+            'created',
+            'modified',
+        ]
+
+
+class VpnSerializer(BaseSerializer):
+    config = serializers.JSONField(initial={})
+
+    class Meta(BaseMeta):
+        model = Vpn
+        fields = [
+            'id',
+            'name',
+            'host',
+            'organization',
+            'key',
+            'ca',
+            'cert',
+            'backend',
+            'notes',
+            'dh',
+            'config',
+        ]
+
+
+class BaseJsonField(serializers.Field):
+    def to_internal_value(self, data):
+        if type(data) is str:
+            if len(data) < 2:
+                data = '{}'
+            return ast.literal_eval(data)
+        else:
+            return data
+
+
+class JsonConfigField(BaseJsonField):
+    def to_representation(self, value):
+        return json.loads(json.dumps(value))
+
+
+class JsonContextField(BaseJsonField):
+    def to_representation(self, value):
+        return dict(value)
+
+
+class FilterTemplatesByOrganization(serializers.PrimaryKeyRelatedField):
+    def get_queryset(self):
+        user = self.context['request'].user
+        if user.is_superuser:
+            queryset = Template.objects.all()
+        else:
+            org_id = next(iter(user.organizations_dict))
```

Comment:
Yes, you are right. I will try to implement this way.

Reply:
A device can get templates of the same organization or shared templates (with organization=None), templates from multiple organizations cannot be assigned.

We need to be able to ensure a user can assign normal templates from the same org and shared templates, please create a test for this @ManishShah120 because we need to ensure this kind of basic functionality will keep working over subsequent changes of the API.

Reply:
> But, I think a device will belong to only one organization, and nemesisdesign suggested that template of multiple organization can't be allowed to be added to a device. So multiple organization template assigning to a single device will not be allowed.
> It's a bit confusing. 🤔 
> -@ManishShah120 on Gitter

Okay, I misunderstood the function's purpose.
Can we please change it's name to suggest it filter's name by organization for a device?

Now, if the data is filtered for the device, why are we looking at `user.organization_dict`?
User can be in multiple orgs while device can only be in one org.
If we use `user.organization_dict`, how do we know we will get the correct organization of the device?
What do you think about using the organization of the device?

Reply:
> Now, if the data is filtered for the device, why are we looking at `user.organization_dict`?

@atb00ker Since this filter is used to filter out the templates belonging to user's organization(s), so suppose a user wants to create a device, then with the help of this function, all the templates belogning to user's org's will be shown, i.e., even if the device is not created the templates in the `config` section needs to be filtered by the users organization(s), so that a user can't use any templates from another organization when creating a device.

> If we use user.organization_dict, how do we know we will get the correct organization of the device?

Yes, so if incase a user belongs to multiple orgs, then he will be allowed to create devices in any one of the orgs he is a part of, and then he/ she should select those organization templates which belongs to the one selected in the device organization field.

Reply:
Okay, let's see this part in Admin UI:
When we create a device, it loads templates *after* we select an organization for device.
We need to do something similar although, I am not sure if it's possible in RDF to dynamically update the multiselect options based on another dropdown, can you please check this? If you don't think anything, please let me know, I'll try to help with the R&D.

The problem with the solution I proposed is that we would should all the templates to the user and add checks in the backend to not allow some of those to be applicable, which is not ideal, we should not even show the templates that are not applicable.

The problem with `next(iter(user.organizations_dict))` is that it returns only one organization. What if it gets templates for org1 and user selected org2 for the device, again, that's a problem.

So, if we want to select organization for device's organization, I don't think we will find our answer in `user.organization_dict`.

Reply:
yes @atb00ker it is better to go with your suggestion, I understood your point.